Adding TE-18.3: MPLS in UDP Encapsulation Scale Test

keysightgems opened the pull request:
**Readme Location:** https://github.com/openconfig/featureprofiles/blob/main/feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/README.md

**The log is attached here:** https://partnerissuetracker.corp.google.com/issues/415458482

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1824 @@
+	e1.Src().SetValue(fa.srcMac)
+	e1.Dst().SetValue(fa.dstMac)
+
+	// For MPLS-in-UDP testing, we only support IPv6 flows
```
No, covered both V4 and V6 as per the Readme. Removed that comment.
Below is the logs:
   mpls_in_udp_scale_test.go:862: 
        OTG Flow Metrics
        --------------------------------------------------------------------------------
        
        Name                     Frames Tx      Frames Rx      FPS Tx         FPS Rx         
        ip4mpls_p5               150787         150787         0              0              
        ip6mpls_p5               150787         150787         0              0              
        --------------------------------------------------------------------------------
        
        
        
    mpls_in_udp_scale_test.go:862: Flow ip4mpls_p5: OutPkts=150787, InPkts=150787, LossPct=0
    mpls_in_udp_scale_test.go:862: Traffic validation PASSED: Flow ip4mpls_p5 has 0% packet loss
    mpls_in_udp_scale_test.go:862: Load balancing has been verified on the Port interfaces.
    mpls_in_udp_scale_test.go:862: Flow ip6mpls_p5: OutPkts=150787, InPkts=150787, LossPct=0
    mpls_in_udp_scale_test.go:862: Traffic validation PASSED: Flow ip6mpls_p5 has 0% packet loss
    mpls_in_udp_scale_test.go:862: Load balancing has been verified on the Port interfaces.

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1824 @@
+}
+
+// getFlow creates a traffic flow for MPLS-in-UDP testing
+func (fa *flowAttr) getFlow(flowType string, name string, dscp uint32) gosnappi.Flow {
```
updated as per the comment.

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1824 @@
+	}
+}
+
+// sendTraffic sends traffic flows for the specified duration
```
Updated!

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1824 @@
+		inPkts := float32(gnmi.Get(t, otg, gnmi.OTG().Flow(flow.Name()).Counters().InPkts().State()))
+		lossPct := ((outPkts - inPkts) * 100) / outPkts
+
+		t.Logf("Flow %s: OutPkts=%v, InPkts=%v, LossPct=%v", flow.Name(), outPkts, inPkts, lossPct)
+
+		if outPkts == 0 {
+			t.Fatalf("OutPkts for flow %s is 0, want > 0", flow.Name())
+		}
+
+		if match {
+			// Expecting traffic to pass (0% loss)
+			if got := lossPct; got > 0 {
+				t.Fatalf("Traffic validation FAILED: Flow %s has %v%% packet loss, want 0%%", flow.Name(), got)
+			} else {
+				t.Logf("Traffic validation PASSED: Flow %s has 0%% packet loss", flow.Name())
+			}
+		} else {
+			// Expecting traffic to fail (100% loss)
+			if got := lossPct; got != 100 {
+				t.Fatalf("Traffic validation FAILED: Flow %s has %v%% packet loss, want 100%%", flow.Name(), got)
+			} else {
+				t.Logf("Traffic validation PASSED: Flow %s has 100%% packet loss", flow.Name())
+			}
+		}
+		if match {
+			rxPorts := []string{ateConfig.Ports().Items()[2].Name(), ateConfig.Ports().Items()[3].Name()}
+			weights := testLoadBalance(t, ate, rxPorts, flow)
+			loadBalVal := true
+			for idx, weight := range portsTrafficDistribution {
+				if got, want := weights[idx], weight; got < (want-tolerance) || got > (want+tolerance) {
+					t.Errorf("ECMP Percentage for Aggregate Index: %d: got %d, want %d", idx+1, got, want)
+					loadBalVal = false
+				}
+			}
+			if loadBalVal {
+				t.Log("Load balancing has been verified on the Port interfaces.")
+			}
+		}
+	}
+}
```
Updated as per the comment.

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1824 @@
+	outerDSCP uint8,
+) []fluent.GRIBIEntry {
+
+	totalEntries := len(vrfs) * (1 + numNHGs + numNHGs) // NH + NHGs + routes
+	entries := make([]fluent.GRIBIEntry, 0, totalEntries)
```
updated.

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1824 @@
+	return entries
+}
+
+func programMPLSinUDPMultiEntries(
+	t *testing.T,
+	dut *ondatra.DUTDevice,
+	vrfs []string,
+	mplsNHBase uint64,
+	mplsLabelStart uint64,
+	numNHGs int,
+	outerIPv6Src, outerIPv6Dst string,
+	outerDstUDPPort uint16,
+	outerIPTTL uint8,
+	outerDSCP uint8,
+) []fluent.GRIBIEntry {
```
Modified.

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1824 @@
+	applyForwardingPolicy(t, dp1.Name())
+	applyForwardingPolicy(t, dp2.Name())
+	// Set static ARP for gRIBI NH MAC resolution
+	if deviations.GRIBIMACOverrideWithStaticARP(dut) {
+		staticARPWithSecondaryIP(t, dut)
+	} else if deviations.GRIBIMACOverrideStaticARPStaticRoute(dut) {
+		staticARPWithMagicUniversalIP(t, dut)
```
updated!

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+			),
+	}
+	// Create NextHopGroups pointing to the same NH
+	for i := 0; i < numNHGs; i++ {
```
Since `i` is starting from 0, you can use the condensed format here.

https://go.dev/play/p/1vjrIWQpx0f

```suggestion
	for i := range numNHGs {
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+		)
+
+		// NextHopGroups and Routes
+		for i := 0; i < numNHGs; i++ {
```
Since i is starting from 0, you can use the condensed format here.


```suggestion
		for i := range numNHGs {
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+}
+
+// validateMPLSPacketCapture validates MPLS-in-UDP encapsulated packets from capture
+func validateMPLSPacketCapture(t *testing.T, ate *ondatra.ATEDevice, otgPortName string, pr *packetResult, labelList []uint64) {
```
https://google.github.io/styleguide/go/best-practices#leave-testing-to-the-test-function

This func follows the ***assertion helper*** pattern, which is not idiomatic in Go.

To fix, have the func return an error instead of using `t.Effort`/`t.Fatal` here.  

The errors should be plumbed back to the main `TestXXX` func, where you`t.Error`/`t.Fatal` should be used.

You will need to adjust the error message formatting per https://google.github.io/styleguide/go/decisions#error-strings

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+	t.Logf("UDP-IPv6 packets found: %d", mplsPacketCount)
+	t.Logf("Valid MPLS-in-UDP packets: %d", validMplsPacketCount)
+
+	if packetCount == 0 {
+		t.Errorf("No packets captured on port %s", otgPortName)
+	} else if mplsPacketCount == 0 {
+		t.Errorf("No UDP-IPv6 packets found in capture on port %s", otgPortName)
+	} else if validMplsPacketCount == 0 {
+		t.Errorf("No valid MPLS-in-UDP packets found in capture on port %s", otgPortName)
+	} else if validMplsPacketCount < (mplsPacketCount / 2) {
+		t.Errorf("Many packets (%d/%d) failed validation", mplsPacketCount-validMplsPacketCount, mplsPacketCount)
+	} else {
+		t.Logf("Packet capture validation PASSED: Found %d valid MPLS-in-UDP packets", validMplsPacketCount)
+	}
```
A `switch` is more idiomatic in Go - https://go.dev/doc/effective_go#switch


```suggestion
	switch {
	case packetCount == 0:
		t.Errorf("No packets captured on port %s", otgPortName)
	case mplsPacketCount == 0:
		t.Errorf("No UDP-IPv6 packets found in capture on port %s", otgPortName)
	case validMplsPacketCount == 0:
		t.Errorf("No valid MPLS-in-UDP packets found in capture on port %s", otgPortName)
	case validMplsPacketCount < (mplsPacketCount / 2):
		t.Errorf("Many packets (%d/%d) failed validation", mplsPacketCount-validMplsPacketCount, mplsPacketCount)
	default:
		t.Logf("Packet capture validation PASSED: Found %d valid MPLS-in-UDP packets", validMplsPacketCount)
	}
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+// configureVrfProfiles implements the “Single/Multi VRF Validation” for Profile 1 (baseline) and Profile 4 (ECMP).
+// It programs MPLS-in-UDP NHs, NHGs, and 20k prefixes (10k v4 + 10k v6),
+// validates FIB/AFT, sends traffic, checks MPLS-over-UDP encapsulation, and deletes entries.
+func configureVrfProfiles(
+	t *testing.T,
+	ateConfig gosnappi.Config,
+	ctx context.Context,
+	dut *ondatra.DUTDevice,
+	ate *ondatra.ATEDevice,
+	c *gribi.Client,
+	vrfs []string,
+	profile int, // 1 = baseline, 4 = ECMP
+) {
```
Single line please.

From https://google.github.io/styleguide/go/guide#line-length:
>Do not split a line:
> * Before an [indentation change](https://google.github.io/styleguide/go/decisions#indentation-confusion) (e.g., function declaration, conditional)

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+	totalNHGs := cfg.totalNHGs
+	nhsPerNHG := cfg.nhsPerNHG
+	totalNHs := totalNHGs * nhsPerNHG
+	var entries []fluent.GRIBIEntry
+	var wantAdds []*client.OpResult
+	var wantDels []*client.OpResult
+	var flows []gosnappi.Flow
+	var labelList []uint64
```
Use a common `var()` block.


```suggestion
	var (
		entries []fluent.GRIBIEntry
		wantAdds []*client.OpResult
		wantDels []*client.OpResult
		flows []gosnappi.Flow
		labelList []uint64
	)
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1470 @@
+		dutIPv6 := fmt.Sprintf(prefixFmtV6, (5*(i+1))+2)
+		ateIPv6 := fmt.Sprintf(prefixFmtV6, (5*(i+1))+1)
+		name := fmt.Sprintf("%s%d", Name, i)
+		mac, _ := incrementMAC(Mac, i+1)
```
https://google.github.io/styleguide/go/decisions#handle-errors
Errors need to be handled in some fashion, even if that is just to log a warning.

cpubob commented on internal/iputil/iputil_test.go:
```diff
@@ -43,3 +44,65 @@ func TestGenerateIPs(t *testing.T) {
 }
+
+// TestGenerateIPv6Prefix covers valid and invalid cases for IPv6 prefix generation.
+func TestGenerateIPv6Prefix(t *testing.T) {
```
Please add test cases for ipv4 IP as input

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1470 @@
+	return rxs
+}
+
+// incrementMAC increments the MAC by i. Returns error if the mac cannot be parsed or overflows the mac address space.
```
Similar functions are being added into the iputil lib as part of PR 4443.  

https://github.com/openconfig/featureprofiles/pull/4443/files

Please rebase onto that PR and use the standard util funcs.

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1470 @@
+		labelList []uint64
+	)
+	// === Program MPLS-in-UDP NH & NHG entries ===
+	switch profile {
```
Please define an enum for the profile so the values have meaning.

```
type vrfProfile int

const (
  profileUnknown vrfProfile = iota
  profileMultiEntries
  profileMultiOpResult
  profileUDPSkewed
  ....

)
```

Then reference those profiles here.

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1470 @@
+}
+
+// configureVrfProfiles implements the “Single/Multi VRF Validation” for Profile 1 (baseline) and Profile 4 (ECMP). It programs MPLS-in-UDP NHs, NHGs, and 20k prefixes (10k v4 + 10k v6), validates FIB/AFT, sends traffic, checks MPLS-over-UDP encapsulation, and deletes entries.
+func configureVrfProfiles(ctx context.Context, t *testing.T, ateConfig gosnappi.Config, dut *ondatra.DUTDevice, ate *ondatra.ATEDevice, c *gribi.Client, vrfs []string, profile int, otgMultiPortCaptureSupported bool) error {
```
https://google.github.io/styleguide/go/decisions#initialisms

`Vrf` -> `VRF`

Please update comment also.

```suggestion
func configureVRFProfiles(ctx context.Context, t *testing.T, ateConfig gosnappi.Config, dut *ondatra.DUTDevice, ate *ondatra.ATEDevice, c *gribi.Client, vrfs []string, profile int, otgMultiPortCaptureSupported bool) error {
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1470 @@
+	t.Helper()
+	dut := ondatra.DUT(t, "dut")
+	ate := ondatra.ATE(t, "ate")
+	ctx := context.Background()
```
Please use `t.Context()` inside tests.

https://google.github.io/styleguide/go/decisions#contexts

```suggestion
	ctx := t.Context()
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1470 @@
+	case validMplsPacketCount == 0:
+		return fmt.Errorf("no valid MPLS-in-UDP packets found on port %s", otgPortName)
+	case validMplsPacketCount < mplsPacketCount/2:
+		return fmt.Errorf("too many invalid packets: %d/%d",
+			mplsPacketCount-validMplsPacketCount, mplsPacketCount)
```
Single line please.

https://google.github.io/styleguide/go/guide#line-length

henryhu-G reviewed (commented):
I'll let Ashwin to continue the review

keysightgems commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -907,3 +920,55 @@ func ConfigureDutWithGueDecap(t *testing.T, dut *ondatra.DUTDevice, guePort int,
+
+		rule := policy.GetOrCreateRule(ruleID)
+		rule.GetOrCreateIpv6().SourceAddress = ygot.String(fmt.Sprintf(cfg.PBFIPv6, i))
+		rule.GetOrCreateIpv6().DscpSet = []uint8{uint8(i % 64)}
```
Yes required, as per the Readme need different DSCPs.

abansod-dev commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1448 @@
+		index := uint32(i + 1)
+		vlanID := uint16(startVLANPort + i)
+		if deviations.NoMixOfTaggedAndUntaggedSubinterfaces(dut) {
+			vlanID += 1
```
Can we add  a comment to explain why we need to increment the vlanID in this case ?

abansod-dev commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1448 @@
+	eth.Ipv6Addresses().Add().SetName(Name + ".IPv6").SetAddress(ateIPv6).SetGateway(dutIPv6).SetPrefix(uint32(ipv6PrefixLen))
+}
+
+// configureATESubinterfaces configures 16 ATE subinterfaces on the target device It returns a slice of the corresponding ATE IPAddresses.
```
nit: Missing period
https://google.github.io/styleguide/go/decisions#comment-sentences

configureATESubinterfaces configures 16 ATE subinterfaces on the target device. It returns a slice of the corresponding ATE IPAddresses.

abansod-dev commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1448 @@
+		return fmt.Errorf("no UDP-IPv6 packets found on port %s", otgPortName)
+	case validMplsPacketCount == 0:
+		return fmt.Errorf("no valid MPLS-in-UDP packets found on port %s", otgPortName)
+	case validMplsPacketCount < mplsPacketCount/2:
```
Do we expect close to half the mpls packets to fail the mpls validation ? 
Why are we logging error only when validMplsPacketCount is less than half of mplsPacketCount ?

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1448 @@
+		prefix := fmt.Sprintf("%s%s", net.IP(ipBytes), routeV6PrfLen)
+
+		// Add IPv6 entry
+		entries = append(entries, fluent.IPv6Entry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).WithPrefix(prefix).WithNextHopGroup(nhgID).WithNextHopGroupNetworkInstance(deviations.DefaultNetworkInstance(dut)))
```
The default VRF is configured here as specified in the README. Non-default VRFs required for multi-VRF test cases are handled in the **programMPLSinUDPMultiEntries** method.

abansod-dev replied:
Thanks we will update the README to reflect the changes I was suggesting. 
We expect IP Entries to always be in non-default VRF. NH and NHG will be in default VRF.

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1448 @@
+	c.FlushAll(t)
+	programBasicEntries(t, dut, &c)
+	t.Run("Profile-1-Single VRF", func(t *testing.T) {
+		if err := configureVRFProfiles(ctx, t, ateConfig, dut, ate, &c, []string{deviations.DefaultNetworkInstance(dut)}, profileSingleVRF, false); err != nil {
```
As per the readme added default VRF, is it required to update? if yes need to update readme as well right?

abansod-dev replied:
Yes. We are working on updating the  README.

ksgireesha commented:
@abansod-dev  has filed a bug [b/450015270](b/450015270) and created PR #4683 to update the README for single VRF configurations. The update clarifies that IP entries must be in a non-default VRF, while the Next-Hop (NH) and Next-Hop Group (NHG) must be in the default VRF. This is necessary to facilitate a change for Keysight in PR #4515. The new PR is currently awaiting approval.

cpubob commented on internal/cfgplugins/dut_initialize.go:
```diff
@@ -16,17 +16,28 @@ package cfgplugins
+type VRFConfig struct {
+	VRFCount      int
+	EnablePBF     bool
+	VrfPolicyName string
+	VrfIPv6       string
```
Please maintain consistency with initialisms.

`Vrf` -> `VRF`

https://google.github.io/styleguide/go/decisions#initialisms

```suggestion
	VRFPolicyName string
	VRFIPv6       string
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1458 @@
+// mustConfigureATESubinterfaces configures 16 ATE subinterfaces on the target device It returns a slice of the corresponding ATE IPAddresses.
+func mustConfigureATESubinterfaces(t *testing.T, ateConfig gosnappi.Config, atePort *ondatra.Port, dut *ondatra.DUTDevice, Name, Mac, prefixFmtV4, prefixFmtV6 string, startVLANPort int) {
+	t.Helper()
+	for i := 0; i < vlanCount; i++ {
```
Since your counter starts are 0, you can use the more simplified range format.

Working example: https://go.dev/play/p/RVk2HFQ6QJ_z

```suggestion
	for i := range vlanCount {
```

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1458 @@
+			),
+	)
+
+	// Pre-parse base IP once
+	ipInt := big.NewInt(0).SetBytes(net.ParseIP(baseIPv6).To16())
+	tmpBytes := make([]byte, 16)
+
+	// Create NextHopGroups and IPv6 routes
+	for i := range numNHGs {
+		nhgID := nextHopGroupID + uint64(i)
+
+		// Add NHG
+		entries = append(entries, fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).WithID(nhgID).AddNextHop(nextHopID, 1))
+
+		// Compute next prefix
+		ipInt.Add(ipInt, big.NewInt(1))
+		ipBytes := ipInt.FillBytes(tmpBytes)
+		prefix := fmt.Sprintf("%s%s", net.IP(ipBytes), routeV6PrfLen)
```
This logic overlaps the helper you created in the iputil library.  Delete here and just use helper.


```suggestion
	// Pre-parse base IP once
	baseIPv6 := net.ParseIP(baseIPv6).To16()
	
	// Create NextHopGroups and IPv6 routes
	for i := range numNHGs {
		nhgID := nextHopGroupID + uint64(i)

		// Add NHG
		entries = append(entries, fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).WithID(nhgID).AddNextHop(nextHopID, 1))

		// Compute next prefix
		offset := iputil.GenerateIPv6WithOffset(baseIPv6, i)
		prefix := fmt.Sprintf("%s/%s", offset.String(), routeV6PrfLen)
```

abansod-dev commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1452 @@
+				prefix := fmt.Sprintf("%s%s", prefixIP.String(), routeV4PrfLen)
+				nhgID := nhgBase + uint64(i%num)
+
+				entries[i] = fluent.IPv4Entry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).WithPrefix(prefix).WithNextHopGroup(nhgID).WithNextHopGroupNetworkInstance(deviations.DefaultNetworkInstance(dut))
```
Similarly this.

keysightgems commented:
@abansod-dev
Addressed all the comments and fixed.

Noticed that the traffic is not flowing through the non-default VRF and is always taking the default VRF path. I have raised issue **455792904** and informed the Arista team. I will apply the fix once a solution is provided.

cpubob commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1680 @@
+	labelMap := make(map[uint64]struct{}, len(labelList))
+	for _, l := range labelList {
+		labelMap[l] = struct{}{}
+	}
```
Using `bool` instead of an empty struct is more common and reads much cleaner.

Also, please don't repeat var type in name - https://google.github.io/styleguide/go/decisions#variable-name-vs-type

```suggestion
    seenLabels := make(map[uint64]bool, len(labelList))
	for _, l := range labelList {
		seenLabels[l] = true
	}
```

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1665 @@
+		// Single NextHop shared by all VRFs:
+		entries = append(entries,
+			fluent.NextHopEntry().
+				WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
```
Placing the entries = append(entries, fluent.NextHopEntry doesn't look good to me: the GRIBI usually expects the next hop ID to be unique, and if you place it here, it will create duplicated nexthops with the same ID. You should move the nexthopentry creation outside the VRF loop

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1665 @@
+}
+
+// buildIPv6Routes generates num IPv6 prefixes per VRF and maps to nhgBase + vrfIdx*num + i.
+func buildIPv6Routes(t *testing.T, dut *ondatra.DUTDevice, num int, baseIPv6 string, vrfs []string, nhgBase uint64) []fluent.GRIBIEntry {
```
Route generation logic is duplicated: programMPLSinUDPEntries was trying to append route entries, and configureVRFProfiles was also calling buildIPv6Routes.

You need to define the scope of the functionalities of each function, e.g.,: programMPLSinUDPEntries should only create NH and NHGs, buildIPv6Routes should only create routes and don't touch NH and NHGs, etc.

keysightgems commented:
@tengyiG 
Addressed all the comments and fixed.

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1784 @@
+	}
+	tDstLists   = []string{}
+	labelLists  = []uint64{}
+	pfx1V6Lists = []string{}
```
Unused. Delete

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1784 @@
+		}
+	})
+	t.Run("Profile-2-Multi VRF", func(t *testing.T) {
+		if err := configureVRFProfiles(ctx, t, ateConfig, dut, ate, vrfsList, profileMultiVRF, true, dstMac); err != nil {
```
These are run sequentially: so the first failure would cause ALL subsequent profiles problematic. Consider t.Fatalf

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1784 @@
+		c.AddEntries(t, batch, nil)
+	}
+	routeTime := time.Now()
+	for time.Since(routeTime) < batchTimeout {
```
add sleep interval to avoid gNMI request flooding

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1784 @@
+			case deviations.GRIBIMACOverrideStaticARPStaticRoute(dut):
+				nh, _ := gribi.NHEntry(nhIDs[i], "MACwithInterface", defaultNI, fluent.InstalledInFIB, &gribi.NHOptions{Interface: port.Name(), Mac: dstMac, Dest: dstIP})
+				allEntries = append(allEntries, nh)
+				// ops = append(ops, op)
```
commented out

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1784 @@
+// expandVRFsBySkew expand VRFs using skewPattern
+func expandVRFsBySkew(vrfs []string, skewPattern []int) []string {
+	if len(vrfs) != len(skewPattern) {
+		panic("vrfs and skewPattern length mismatch")
```
don't panic, return an error

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1784 @@
+		srcPort:  []string{"port1", "port2"},
+		dstPorts: []string{"port3", "port4"},
+		topo:     gosnappi.NewConfig(),
+	}
```
Never pushed, delete.

tengyiG reviewed (changes requested):
Please check the comments carefully and fix all of them

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+	return entries, vrfV6PfxMap
+}
+
+// programMPLSinUDPMultiEntries programs MPLS-in-UDP encapsulated NextHops, NextHopGroups, and IPv6 routes across multiple VRFs. It returns a slice of gRIBI entries that can be pushed to the DUT using the fluent gRIBI client.
```
The NHG entries reference the package-level constant nextHopID instead of the nhID parameter. Same issue in expectedMPLSinUDPMultiOpResults with mplsNHBase.

Recommendation: Replace nextHopID with nhID inside the function body. Apply the same fix to expectedMPLSinUDPMultiOpResults for mplsNHBase.

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+	}
+	tDstLists   = []string{}
+	labelLists  = []uint64{}
+	pfx1V6Lists = []string{}
```
Package-level mutable state never reset between subtests

Recommendation: Either:
Reset all three slices at the top of configureVRFProfiles, or
Eliminate the globals by threading these values through function parameters and return values.

tengyiG commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+	}
+
+	// Define heavier VRFs: e.g., top 3 get more
+	heavyCount := int(math.Max(1, float64(numVRFs)/4)) // first 25% VRFs heavier
```
Use the max builtin, don't have to use math.Max.

tengyiG reviewed (changes requested):
Please take a look at these and fix

keysightgems commented on feature/gribi/mpls_in_udp/otg_tests/mpls_in_udp_scale/mpls_in_udp_scale_test.go:
```diff
@@ -0,0 +1,1757 @@
+	switch profile {
+	case profileMultiVRF:
+		entries, vrfPfxMap = programMPLSinUDPMultiEntries(t, dut, vrfs, nextHopID, nextHopGroupID, mplsLabel, totalNHGs, outerIPv6Src, outerIPv6Dst, outerDstUDPPort, outerTTL, outerDSCP)
+		wantAdds, wantDels = expectedMPLSinUDPMultiOpResults(t, vrfs, nextHopID, nextHopGroupID, totalNHGs)
```
Currently the test invokes c.AddEntries(t, batch, nil), so the Add RPC is not asserting individual operation results.

I tried switching to c.AddEntries(t, batch, testCaseArgs.wantAddResults) to assert expected results explicitly. However, at larger scale this leads to DeadlineExceeded, likely due to the time taken for the DUT to ACK all entries.

ksgireesha commented:
Hi @dplore ,

We have 2 LGTMs for this PR. Please review, run this test in our environment and if it passes, please merge
